Adding SES Api Call caching functionality.

BuggerSee opened the pull request:
Looping and consuming messages with a low enough **--time-limit** (e.g. 1) in a custom mautic cronjob script will result in excess SES api calls. This will result in a **TooManyRequestsException** from amazon ses api. 

Recently I was debugging some strange behaviour. After running some campaigns I was not able to query `aws --region eu-west-1 sesv2 get-account` as I was being rate limited. I don't think this should affect email sending functionality because they are two different API request types. Right?

**But** spamming the api and receiving the **TooManyRequestsException** can have a performance impact, because it will always defaults your maxSendRate to 14, even though your limit may be much higher.

Wouldn't it make sense to cache the maxSendRate from the request for a specific time(e.g. 60 minutes) so the Account Api doesnt get rate limited and you still send full speed? Or is this user error and you should not run the consume message command with such a low time-limit?


I did some modifications to the AmazonSesTransportFactory.php and it seems to work on my setup. 

I will attach some logs of my debugging. 

With caching functionality:
![Bildschirmfoto 2025-02-11 um 19 12 24](https://github.com/user-attachments/assets/2b4305f0-7f90-4a80-8036-bc7a0b7e5e9c)
No caching:
![Bildschirmfoto 2025-02-11 um 19 11 19](https://github.com/user-attachments/assets/9a7f11fe-de26-453b-a171-848c4fc0566a)
No caching results in defaulting maxSendRate to 14.
![Bildschirmfoto 2025-02-11 um 19 10 31](https://github.com/user-attachments/assets/6ad10c4f-845d-4b37-bda9-0afa6524abc9)

pm-pmaas commented:
> Looping and consuming messages with a low enough **--time-limit** (e.g. 1) in a custom mautic cronjob script will result in excess SES api calls. This will result in a **TooManyRequestsException** from amazon ses api.
> 
> Recently I was debugging some strange behaviour. After running some campaigns I was not able to query `aws --region eu-west-1 sesv2 get-account` as I was being rate limited. I don't think this should affect email sending functionality because they are two different API request types. Right?
> 
> **But** spamming the api and receiving the **TooManyRequestsException** can have a performance impact, because it will always defaults your maxSendRate to 14, even though your limit may be much higher.
> 
> Wouldn't it make sense to cache the maxSendRate from the request for a specific time(e.g. 60 minutes) so the Account Api doesnt get rate limited and you still send full speed? Or is this user error and you should not run the consume message command with such a low time-limit?
> 
> I did some modifications to the AmazonSesTransportFactory.php and it seems to work on my setup.
> 
> I will attach some logs of my debugging.
> 
> With caching functionality: ![Bildschirmfoto 2025-02-11 um 19 12 24](https://private-user-images.githubusercontent.com/36350532/412111178-2b4305f0-7f90-4a80-8036-bc7a0b7e5e9c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkzMDEyMjAsIm5iZiI6MTczOTMwMDkyMCwicGF0aCI6Ii8zNjM1MDUzMi80MTIxMTExNzgtMmI0MzA1ZjAtN2Y5MC00YTgwLTgwMzYtYmM3YTBiN2U1ZTljLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjExVDE5MDg0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAzYmFkNzA5YTBkODc5MjA5ZjE5MmJmYjY1NGIxYzEzNjc3MzBkYTZlZGU5ZmIzZmU4ZTQ5YzZiZTk5YWJjM2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2BtPkgPs30RPPAKXRStLKo0ur3qWkdQ1Pa16GDs5rYo) No caching: ![Bildschirmfoto 2025-02-11 um 19 11 19](https://private-user-images.githubusercontent.com/36350532/412111185-9a7f11fe-de26-453b-a171-848c4fc0566a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkzMDEyMjAsIm5iZiI6MTczOTMwMDkyMCwicGF0aCI6Ii8zNjM1MDUzMi80MTIxMTExODUtOWE3ZjExZmUtZGUyNi00NTNiLWExNzEtODQ4YzRmYzA1NjZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjExVDE5MDg0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRlZmU5ZWM2OTFkYzE1NDFmYTMyY2ZiYjg2ZjYxODI0OGVjN2NmMzJhMzVmM2FmMmIxMTA4MDNlZmQxZTc3ZjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qb7D258aGvoXyGldkTrHZSOYLxvBGpND3mVnHyohkwA) No caching results in defaulting maxSendRate to 14. ![Bildschirmfoto 2025-02-11 um 19 10 31](https://private-user-images.githubusercontent.com/36350532/412111187-6ad10c4f-845d-4b37-bda9-0afa6524abc9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkzMDEyMjAsIm5iZiI6MTczOTMwMDkyMCwicGF0aCI6Ii8zNjM1MDUzMi80MTIxMTExODctNmFkMTBjNGYtODQ1ZC00YjM3LWJkYTktMGFmYTY1MjRhYmM5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjExVDE5MDg0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBkZDA3YjA3MWNlNTJhOGYyY2YwNmE3ZWIzZTYyZjhmMWYwZTQwZTUyNzEyZmJkODczYTZlMjJlYWY4NzFmMjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-OxDlLK43LqgqjOm66v2NFHQxDfDMK1iZjY9-x2Y310)

Great idea! Is it possible to extract checking the caching status to a separate function , which make the code a bit more cleaner ? I also suggest to remove or make optional the logging feauture.

After that changes, i will check and merge your PR.

BuggerSee commented:
> > Looping and consuming messages with a low enough **--time-limit** (e.g. 1) in a custom mautic cronjob script will result in excess SES api calls. This will result in a **TooManyRequestsException** from amazon ses api.
> > Recently I was debugging some strange behaviour. After running some campaigns I was not able to query `aws --region eu-west-1 sesv2 get-account` as I was being rate limited. I don't think this should affect email sending functionality because they are two different API request types. Right?
> > **But** spamming the api and receiving the **TooManyRequestsException** can have a performance impact, because it will always defaults your maxSendRate to 14, even though your limit may be much higher.
> > Wouldn't it make sense to cache the maxSendRate from the request for a specific time(e.g. 60 minutes) so the Account Api doesnt get rate limited and you still send full speed? Or is this user error and you should not run the consume message command with such a low time-limit?
> > I did some modifications to the AmazonSesTransportFactory.php and it seems to work on my setup.
> > I will attach some logs of my debugging.
> > With caching functionality: ![Bildschirmfoto 2025-02-11 um 19 12 24](https://private-user-images.githubusercontent.com/36350532/412111178-2b4305f0-7f90-4a80-8036-bc7a0b7e5e9c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkzMDEyMjAsIm5iZiI6MTczOTMwMDkyMCwicGF0aCI6Ii8zNjM1MDUzMi80MTIxMTExNzgtMmI0MzA1ZjAtN2Y5MC00YTgwLTgwMzYtYmM3YTBiN2U1ZTljLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjExVDE5MDg0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAzYmFkNzA5YTBkODc5MjA5ZjE5MmJmYjY1NGIxYzEzNjc3MzBkYTZlZGU5ZmIzZmU4ZTQ5YzZiZTk5YWJjM2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2BtPkgPs30RPPAKXRStLKo0ur3qWkdQ1Pa16GDs5rYo) No caching: ![Bildschirmfoto 2025-02-11 um 19 11 19](https://private-user-images.githubusercontent.com/36350532/412111185-9a7f11fe-de26-453b-a171-848c4fc0566a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkzMDEyMjAsIm5iZiI6MTczOTMwMDkyMCwicGF0aCI6Ii8zNjM1MDUzMi80MTIxMTExODUtOWE3ZjExZmUtZGUyNi00NTNiLWExNzEtODQ4YzRmYzA1NjZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjExVDE5MDg0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRlZmU5ZWM2OTFkYzE1NDFmYTMyY2ZiYjg2ZjYxODI0OGVjN2NmMzJhMzVmM2FmMmIxMTA4MDNlZmQxZTc3ZjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qb7D258aGvoXyGldkTrHZSOYLxvBGpND3mVnHyohkwA) No caching results in defaulting maxSendRate to 14. ![Bildschirmfoto 2025-02-11 um 19 10 31](https://private-user-images.githubusercontent.com/36350532/412111187-6ad10c4f-845d-4b37-bda9-0afa6524abc9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkzMDEyMjAsIm5iZiI6MTczOTMwMDkyMCwicGF0aCI6Ii8zNjM1MDUzMi80MTIxMTExODctNmFkMTBjNGYtODQ1ZC00YjM3LWJkYTktMGFmYTY1MjRhYmM5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjExVDE5MDg0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBkZDA3YjA3MWNlNTJhOGYyY2YwNmE3ZWIzZTYyZjhmMWYwZTQwZTUyNzEyZmJkODczYTZlMjJlYWY4NzFmMjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-OxDlLK43LqgqjOm66v2NFHQxDfDMK1iZjY9-x2Y310)
> 
> Great idea! Is it possible to extract checking the caching status to a separate function , which make the code a bit more cleaner ? I also suggest to remove or make optional the logging feauture.
> 
> After that changes, i will check and merge your PR.

Added the caching logic in separate functions and removed some logging features. Also tested and works on my system.
